Auto register maps from domain creators

nilsdeppe opened the pull request:
## Proposed changes

- Have domain creators specify which maps they will use
- Use the maps info from the domain creators to automatically register the appropriate maps with Charm++ for serialization.

Depends on:
- [x] #2027

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Reduce test cases in FrustalCloak test

nilsdeppe committed:
Have domain creators specify the maps they need registered

Previously we had all the CoordinateMap<...> classes that need to be registered
with Charm listed in RegisterDerivedClassesWithCharm and manually maintained
that file. It's quite easy to have each domain creator list what maps it will
use and then register them automatically with Charm.

nilsdeppe committed:
Register moving mesh maps from TimeDependences' maps_list

kidder commented on src/Domain/Creators/AlignedLattice.hpp:
```diff
@@ -31,6 +46,18 @@ namespace creators {
 template <size_t VolumeDim>
 class AlignedLattice : public DomainCreator<VolumeDim> {
  public:
+  using maps_list = tmpl::list<
```
shouldn't this list depend upon VolumeDim?

nilsdeppe replied:
It could, but as far as I can tell registering something twice doesn't seem to matter. I can double check with the Charm folks. My reason for doing it this way was to avoid a double nested `tmpl::conditional_t` because I found that a bit ugly.

nilsdeppe replied:
One fairly easy way to get around double registration is changing the way the registration function works. I'll do that and then we don't have to worry about double registering anything.

kidder commented on src/Domain/Creators/TimeDependence/UniformTranslation.hpp:
```diff
@@ -57,6 +60,15 @@ class UniformTranslation final : public TimeDependence<MeshDim> {
   using Translation = domain::CoordMapsTimeDependent::Translation;
 
  public:
+  using maps_list = tmpl::list<
```
again, shouldn't this list depend upon MeshDim?

kidder commented on src/Domain/Creators/AlignedLattice.cpp:
```diff
@@ -3,6 +3,10 @@
 
 #include "Domain/Creators/AlignedLattice.hpp"
 
+#include "Domain/CoordinateMaps/CoordinateMap.hpp"
```
I'm a bit puzzled by all the extra includes needed in this commit; I don't see what they have to do with registration of maps_list of TimeDependences

nilsdeppe replied:
I needed them in the tests to get linking working with GCC7 release. I'm of the opinion that the domain creators should include the coordinate maps that they use, but I'm fine removing them if you prefer.

nilsdeppe replied:
Thinking more about this, I'm confused that the required code isn't instantiated anywhere. I'll think about this a bit more and try to figure it out, then report back.

nilsdeppe replied:
Okay, I looked at this quite a bit and I half understand what's going on.
- The includes in the non-test files are not needed, I've removed them in a fixup
- The includes in the test files are needed for a few different reasons (though I don't understand them all).
  1. The compiler needs to generate instantiations for calling the `_impl` member functions of `CoordinateMap`. I don't know why the linker can't find them elsewhere when the main executables seem to be fine. 
  2. The tests pass one of the maps by rvalue reference and the other by const lvalue reference to `make_coordinate_map_base`, while the code that creates the maps in the domain creator I think does 2 lvalue or 2 rvalue references (I didn't bother digging through it). This is a legitimate reason to need the `CoordinateMap.tpp` include in the test files.
  3. The linker cannot find pup functions for the `ProductOf?Maps`. Again, I don't understand why the executables are fine but the tests aren't. This makes no sense to me.

  For 2., the answer is including the `tpp`. For 1 and 3 I've written some macros to auto-instantiate all the necessary code. I'm not sure of a good way to test that all the explicit instantiations are generated because as I mentioned some includes for `tpp`s are actually needed by the tests. I can add a commit that adds the explicit instantiation macros and uses them for the cases encountered here if desired.

moxcodes commented on tests/Unit/Domain/Creators/Test_Rectangle.cpp:
```diff
@@ -114,22 +119,6 @@ void test_rectangle() {
            {Direction<2>::lower_eta(), {0, aligned_orientation}},
            {Direction<2>::upper_eta(), {0, aligned_orientation}}}},
       std::vector<std::unordered_set<Direction<2>>>{{}});
-
```
`Test_Interval.cpp` has a similar registration and serialization part of the test that seems like it could also be removed.

nilsdeppe commented:
I've pushed fixups and responded to the `include`s comment in as much detail as I can. I'd like some input on what to do with the includes. Thanks!

nilsdeppe committed:
Explicitly instantiate CoordinateMap classes

- Adds MapInstantiationMacros.hpp which provides macros for instantiation all
  the member functions of the CoordinateMap class
- Instantiate coordinate maps in DomainHelpers.hpp

nilsdeppe commented:
I pushed an extra commit that does a bunch of explicit instantiations of member functions for `CoordinateMap` and adds macros to generate the instantiations

kidder commented:
I am happy for you to squash if @moxcodes is

moxcodes commented:
fixups look good, go ahead and squash

nilsdeppe commented:
Rebased on develop and squashed. Thanks for the reviews! :)
